File: common/mobile/modules/blinky-secure-socket/ios/BlinkySecureSocketModule.swift
import CryptoKit
import ExpoModulesCore
import Foundation
import Security

public final class BlinkySecureSocketModule: Module, URLSessionWebSocketDelegate {
  private lazy var session: URLSession = {
    URLSession(configuration: .ephemeral, delegate: self, delegateQueue: nil)
  }()
  private var sockets: [String: URLSessionWebSocketTask] = [:]
  private var pins: [String: String] = [:]
  private let socketStateLock = NSRecursiveLock()
  private var transferPins: [Int: String] = [:]
  private let transferPinsLock = NSLock()
  private let transferChunkSize = 16 * 1024 * 1024

  public func definition() -> ModuleDefinition {
    Name("BlinkySecureSocket")
    Events("onOpen", "onMessage", "onClose", "onError", "onTransferProgress")

    AsyncFunction("connect") { (id: String, urlString: String, pin: String) throws in
      guard let url = URL(string: urlString), url.scheme?.lowercased() == "wss" else {
        throw SecureSocketError("Secure socket requires a wss:// URL")
      }
      guard pin.hasPrefix("sha256/") else {
        throw SecureSocketError("Secure socket requires a sha256/ certificate pin")
      }
      self.socketStateLock.lock()
      defer { self.socketStateLock.unlock() }
      self.closeSocket(id: id)
      let task = self.session.webSocketTask(with: url)
      self.sockets[id] = task
      self.pins[id] = pin
      task.resume()
    }

    AsyncFunction("sendText") { (id: String, data: String) throws in
      guard let socket = self.socket(for: id) else { throw SecureSocketError("Secure socket is not connected: \(id)") }
      socket.send(.string(data)) { error in
        if let error { self.emitError(id: id, message: error.localizedDescription) }
      }
    }

    AsyncFunction("sendBinary") { (id: String, base64Data: String) throws in
      guard let socket = self.socket(for: id), let data = Data(base64Encoded: base64Data) else {
        throw SecureSocketError("Secure socket binary payload is invalid or disconnected: \(id)")
      }
      socket.send(.data(data)) { error in
        if let error { self.emitError(id: id, message: error.localizedDescription) }
      }
    }

    AsyncFunction("close") { (id: String) in
      self.closeSocket(id: id)
    }
    AsyncFunction("getSecureValue") { (key: String) -> String? in
      self.getSecureValue(key: key)
    }
    AsyncFunction("setSecureValue") { (key: String, value: String) throws in
      try self.setSecureValue(key: key, value: value)
    }
    AsyncFunction("deleteSecureValue") { (key: String) throws in
      try self.deleteSecureValue(key: key)
    }
    AsyncFunction("hashFile") { (uri: String) async throws -> [String: Any] in
      try self.hashFile(uri: uri)
    }
    AsyncFunction("uploadFile") { (options: [String: Any]) async throws -> [String: Any] in
      try await self.uploadFile(options: options)
    }
    AsyncFunction("downloadFile") { (options: [String: Any]) async throws -> String in
      try await self.downloadFile(options: options)
    }

    OnDestroy {
      self.socketIDs().forEach { self.closeSocket(id: $0) }
    }
  }

  public func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didOpenWithProtocol protocol: String?
  ) {
    guard let id = id(for: webSocketTask) else { return }
    sendEvent("onOpen", ["id": id])
    receiveNext(id: id, task: webSocketTask)
  }

  public func urlSession(
    _ session: URLSession,
    webSocketTask: URLSessionWebSocketTask,
    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
    reason: Data?
  ) {
    guard let id = id(for: webSocketTask), removeSocket(id: id, ifCurrent: webSocketTask) else { return }
    sendEvent("onClose", ["id": id])
  }

  public func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didReceive challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
    let expectedPin: String?
    if let socket = task as? URLSessionWebSocketTask, let id = id(for: socket) {
      expectedPin = pin(for: id)
    } else {
      expectedPin = transferPin(for: task.taskIdentifier)
    }
    guard let trust = challenge.protectionSpace.serverTrust,
          let expectedPin,
          certificatePin(for: trust) == expectedPin else {
      completionHandler(.cancelAuthenticationChallenge, nil)
      return
    }
    // The pinned certificate is intentionally private/self-signed, so normal
    // system CA validation is not used after the public-key pin matches.
    completionHandler(.useCredential, URLCredential(trust: trust))
  }

  public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
    removeTransferPin(for: task.taskIdentifier)
    guard let socket = task as? URLSessionWebSocketTask, let id = id(for: socket), let error else { return }
    emitError(id: id, message: error.localizedDescription)
  }

  private func hashFile(uri: String) throws -> [String: Any] {
    let url = try fileURL(uri)
    let hasSecurityScope = url.startAccessingSecurityScopedResource()
    defer { if hasSecurityScope { url.stopAccessingSecurityScopedResource() } }
    let handle = try FileHandle(forReadingFrom: url)
    defer { try? handle.close() }
    var hasher = SHA256()
    var size: UInt64 = 0
    while true {
      guard let data = try handle.read(upToCount: 1024 * 1024), !data.isEmpty else { break }
      hasher.update(data: data)
      size += UInt64(data.count)
    }
    let digest = hasher.finalize().map { String(format: "%02x", $0) }.joined()
    return ["size": size, "sha256": digest]
  }

  private func uploadFile(options: [String: Any]) async throws -> [String: Any] {
    let source = try fileURL(options.string("sourceUri"))
    let transferId = try options.string("transferId")
    let endpoint = try transferURL(options.string("url"))
    let token = try options.string("token")
    let pin = options["pin"] as? String
    let size = try options.uint64("size")
    let chunkSize = min(max(1, options.int("chunkSize", default: transferChunkSize)), transferChunkSize)
    var offset = try options.uint64("offset", default: 0)
    guard size > 0 && offset <= size else { throw SecureSocketError("Invalid upload size or offset") }

    let hasSecurityScope = source.startAccessingSecurityScopedResource()
    defer { if hasSecurityScope { source.stopAccessingSecurityScopedResource() } }
    while offset < size {
      let handle = try FileHandle(forReadingFrom: source)
      defer { try? handle.close() }
      try handle.seek(toOffset: offset)
      let requestBytes = min(UInt64(chunkSize), size - offset)
      guard let data = try handle.read(upToCount: Int(requestBytes)), !data.isEmpty else {
        throw SecureSocketError("Selected file ended before its reported size")
      }
      var request = URLRequest(url: endpoint, timeoutInterval: 300)
      request.httpMethod = "POST"
      request.httpBody = data
      request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
      request.setValue(String(offset), forHTTPHeaderField: "Upload-Offset")
      request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
      request.setValue(String(data.count), forHTTPHeaderField: "Content-Length")
      let (_, response) = try await send(request, pin: pin)
      let next = response.value(forHTTPHeaderField: "Upload-Offset").flatMap(UInt64.init)
      if response.statusCode == 409, let next, next <= size {
        offset = next
        continue
      }
      guard response.statusCode == 200 || response.statusCode == 201 else {
        throw SecureSocketError("Upload failed (HTTP \(response.statusCode))")
      }
      let updated = next ?? offset + UInt64(data.count)
      guard updated > offset && updated <= size else { throw SecureSocketError("PC returned an invalid upload offset") }
      offset = updated
      sendEvent("onTransferProgress", ["id": transferId, "direction": "upload", "bytes": offset, "total": size])
    }
    return ["transferId": transferId, "uploadOffset": offset, "complete": true]
  }

  private func downloadFile(options: [String: Any]) async throws -> String {
    let transferId = try options.string("transferId")
    let baseURL = try transferURL(options.string("url"))
    let token = try options.string("token")
    let pin = options["pin"] as? String
    let size = try options.uint64("size")
    let expectedHash = try options.string("sha256").lowercased()
    let filename = safeName(try options.string("filename"))
    guard size > 0 else { throw SecureSocketError("Invalid download size") }

    let directory = try FileManager.default.url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
      .appendingPathComponent("BlinkyTransfers", isDirectory: true)
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    let partial = directory.appendingPathComponent("\(transferId).part")
    var offset = (try? partial.resourceValues(forKeys: [.fileSizeKey]).fileSize).map(UInt64.init) ?? 0
    if offset > size {
      try FileManager.default.removeItem(at: partial)
      offset = 0
    }

    while offset < size {
      let end = min(size - 1, offset + UInt64(transferChunkSize) - 1)
      var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)!
      components.path = "/download/\(transferId)"
      guard let url = components.url else { throw SecureSocketError("Invalid file download URL") }
      var request = URLRequest(url: url, timeoutInterval: 300)
      request.httpMethod = "GET"
      request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
      request.setValue("bytes=\(offset)-\(end)", forHTTPHeaderField: "Range")
      let (data, response) = try await send(request, pin: pin)
      guard response.statusCode == 206 else { throw SecureSocketError("Download failed (HTTP \(response.statusCode))") }
      guard UInt64(response.value(forHTTPHeaderField: "X-File-Size") ?? "") == size,
            response.value(forHTTPHeaderField: "X-File-Sha256")?.lowercased() == expectedHash else {
        throw SecureSocketError("PC returned different file metadata")
      }
      let expectedChunk = Int(end - offset + 1)
      guard data.count == expectedChunk else { throw SecureSocketError("Download chunk was incomplete") }
      if !FileManager.default.fileExists(atPath: partial.path) {
        FileManager.default.createFile(atPath: partial.path, contents: nil)
      }
      let output = try FileHandle(forWritingTo: partial)
      try output.seekToEnd()
      try output.write(contentsOf: data)
      try output.synchronize()
      try output.close()
      offset += UInt64(data.count)
      sendEvent("onTransferProgress", ["id": transferId, "direction": "download", "bytes": offset, "total": size])
    }

    let localHash = try hashFile(uri: partial.absoluteString)["sha256"] as? String
    guard localHash?.lowercased() == expectedHash else { throw SecureSocketError("Downloaded file SHA-256 did not match the PC") }
    let destination = uniqueTransferURL(directory: directory, filename: filename)
    try FileManager.default.moveItem(at: partial, to: destination)
    return destination.absoluteString
  }

  private func send(_ request: URLRequest, pin: String?) async throws -> (Data, HTTPURLResponse) {
    guard let scheme = request.url?.scheme?.lowercased(), scheme == "http" || scheme == "https" else {
      throw SecureSocketError("Transfer URL must use HTTP or HTTPS")
    }
    if scheme == "https" && (pin?.hasPrefix("sha256/") != true) {
      throw SecureSocketError("HTTPS transfers require the saved certificate pin")
    }
    if scheme == "http" && pin != nil {
      throw SecureSocketError("Pinned release transfers must use HTTPS")
    }
    return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<(Data, HTTPURLResponse), Error>) in
      let task = session.dataTask(with: request) { data, response, error in
        if let error { continuation.resume(throwing: error); return }
        guard let response = response as? HTTPURLResponse else {
          continuation.resume(throwing: SecureSocketError("PC returned an invalid HTTP response")); return
        }
        continuation.resume(returning: (data ?? Data(), response))
      }
      if scheme == "https", let pin { self.setTransferPin(pin, for: task.taskIdentifier) }
      task.resume()
    }
  }

  private func transferPin(for taskIdentifier: Int) -> String? {
    transferPinsLock.lock()
    defer { transferPinsLock.unlock() }
    return transferPins[taskIdentifier]
  }

  private func setTransferPin(_ pin: String, for taskIdentifier: Int) {
    transferPinsLock.lock()
    defer { transferPinsLock.unlock() }
    transferPins[taskIdentifier] = pin
  }

  private func removeTransferPin(for taskIdentifier: Int) {
    transferPinsLock.lock()
    defer { transferPinsLock.unlock() }
    transferPins.removeValue(forKey: taskIdentifier)
  }

  private func fileURL(_ uri: String) throws -> URL {
    guard let url = URL(string: uri), url.isFileURL else { throw SecureSocketError("Selected file URI is invalid") }
    return url
  }

  private func transferURL(_ value: String) throws -> URL {
    guard let url = URL(string: value), let scheme = url.scheme?.lowercased(), scheme == "http" || scheme == "https" else {
      throw SecureSocketError("Transfer URL must use HTTP or HTTPS")
    }
    return url
  }

  private func safeName(_ value: String) -> String {
    let cleaned = value.unicodeScalars.map { scalar -> Character in
      if CharacterSet(charactersIn: "/\\:*?\"<>|").contains(scalar) || CharacterSet.controlCharacters.contains(scalar) { return "_" }
      return Character(scalar)
    }.reduce(into: "") { $0.append($1) }.trimmingCharacters(in: CharacterSet(charactersIn: ". "))
    return cleaned.isEmpty || cleaned == ".." ? "blinky-download.bin" : String(cleaned.prefix(240))
  }

  private func uniqueTransferURL(directory: URL, filename: String) -> URL {
    let original = directory.appendingPathComponent(filename)
    if !FileManager.default.fileExists(atPath: original.path) { return original }
    let stem = original.deletingPathExtension().lastPathComponent
    let ext = original.pathExtension
    for index in 1...99999 {
      let candidateName = ext.isEmpty ? "\(stem) (\(index))" : "\(stem) (\(index)).\(ext)"
      let candidate = directory.appendingPathComponent(candidateName)
      if !FileManager.default.fileExists(atPath: candidate.path) { return candidate }
    }
    return directory.appendingPathComponent("blinky-\(UUID().uuidString)-\(filename)")
  }


  private func receiveNext(id: String, task: URLSessionWebSocketTask) {
    task.receive { [weak self] result in
      guard let self else { return }
      DispatchQueue.main.async {
        guard self.isCurrentSocket(task, for: id) else { return }
        switch result {
        case .success(let message):
          switch message {
          case .string(let text):
            self.sendEvent("onMessage", ["id": id, "data": text])
          case .data(let data):
            self.sendEvent("onMessage", ["id": id, "data": data.base64EncodedString()])
          @unknown default:
            break
          }
          self.receiveNext(id: id, task: task)
        case .failure(let error):
          self.emitError(id: id, message: error.localizedDescription)
        }
      }
    }
  }

  private func closeSocket(id: String) {
    socketStateLock.lock()
    let socket = sockets.removeValue(forKey: id)
    pins.removeValue(forKey: id)
    socketStateLock.unlock()
    socket?.cancel(with: .normalClosure, reason: nil)
  }

  private func id(for task: URLSessionWebSocketTask) -> String? {
    socketStateLock.lock()
    defer { socketStateLock.unlock() }
    return sockets.first(where: { $0.value === task })?.key
  }

  private func socket(for id: String) -> URLSessionWebSocketTask? {
    socketStateLock.lock()
    defer { socketStateLock.unlock() }
    return sockets[id]
  }

  private func pin(for id: String) -> String? {
    socketStateLock.lock()
    defer { socketStateLock.unlock() }
    return pins[id]
  }

  private func socketIDs() -> [String] {
    socketStateLock.lock()
    defer { socketStateLock.unlock() }
    return Array(sockets.keys)
  }

  private func isCurrentSocket(_ task: URLSessionWebSocketTask, for id: String) -> Bool {
    socketStateLock.lock()
    defer { socketStateLock.unlock() }
    return sockets[id] === task
  }

  private func removeSocket(id: String, ifCurrent task: URLSessionWebSocketTask) -> Bool {
    socketStateLock.lock()
    defer { socketStateLock.unlock() }
    guard sockets[id] === task else { return false }
    sockets.removeValue(forKey: id)
    pins.removeValue(forKey: id)
    return true
  }

  private func emitError(id: String, message: String) {
    sendEvent("onError", ["id": id, "error": message])
  }

  private let keychainService = "app.blinky.tutor.remote"

  private func keychainQuery(key: String) -> [String: Any] {
    [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: keychainService,
      kSecAttrAccount as String: key,
    ]
  }

  private func getSecureValue(key: String) -> String? {
    var query = keychainQuery(key: key)
    query[kSecReturnData as String] = true
    query[kSecMatchLimit as String] = kSecMatchLimitOne
    var result: CFTypeRef?
    guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
          let data = result as? Data else { return nil }
    return String(data: data, encoding: .utf8)
  }

  private func setSecureValue(key: String, value: String) throws {
    let data = Data(value.utf8)
    SecItemDelete(keychainQuery(key: key) as CFDictionary)
    var query = keychainQuery(key: key)
    query[kSecValueData as String] = data
    guard SecItemAdd(query as CFDictionary, nil) == errSecSuccess else {
      throw SecureSocketError("Unable to save secure value")
    }
  }

  private func deleteSecureValue(key: String) throws {
    let status = SecItemDelete(keychainQuery(key: key) as CFDictionary)
    guard status == errSecSuccess || status == errSecItemNotFound else {
      throw SecureSocketError("Unable to delete secure value")
    }
  }

  private func certificatePin(for trust: SecTrust) -> String? {
    guard let certificate = SecTrustGetCertificateAtIndex(trust, 0),
          let key = SecCertificateCopyKey(certificate),
          let rawKey = SecKeyCopyExternalRepresentation(key, nil) as Data?,
          rawKey.count == 65 else {
      return nil
    }

    // The desktop identity is P-256. This is the DER SubjectPublicKeyInfo
    // prefix for id-ecPublicKey + prime256v1 followed by the 65-byte key.
    let spkiPrefix = Data([
      0x30, 0x59, 0x30, 0x13, 0x06, 0x07, 0x2A, 0x86, 0x48, 0xCE, 0x3D,
      0x02, 0x01, 0x06, 0x08, 0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x03, 0x01,
      0x07, 0x03, 0x42, 0x00
    ])
    let digest = SHA256.hash(data: spkiPrefix + rawKey)
    return "sha256/" + Data(digest).base64EncodedString()
  }
}

private struct SecureSocketError: Error, LocalizedError {
  let message: String
  init(_ message: String) { self.message = message }
  var errorDescription: String? { message }
}

private extension Dictionary where Key == String, Value == Any {
  func string(_ key: String) throws -> String {
    guard let value = self[key] as? String else { throw SecureSocketError("Missing \(key)") }
    return value
  }

  func uint64(_ key: String, default fallback: UInt64? = nil) throws -> UInt64 {
    guard let value = self[key] else {
      if let fallback { return fallback }
      throw SecureSocketError("Missing \(key)")
    }
    guard let number = value as? NSNumber, number.doubleValue >= 0, number.doubleValue <= Double(UInt64.max) else {
      throw SecureSocketError("Invalid \(key)")
    }
    return number.uint64Value
  }

  func int(_ key: String, default fallback: Int) -> Int {
    (self[key] as? NSNumber)?.intValue ?? fallback
  }
}

File: common/mobile/modules/blinky-secure-socket/android/src/main/java/com/technerds/blinky/securesocket/BlinkySecureSocketModule.kt
package com.technerds.blinky.securesocket

import android.util.Base64
import android.content.Context
import android.net.Uri
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import java.net.URI
import java.net.HttpURLConnection
import java.net.URL
import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.concurrent.ConcurrentHashMap
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.security.KeyStore
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager

class BlinkySecureSocketModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()
  private val sockets = ConcurrentHashMap<String, WebSocket>()
  private val clients = ConcurrentHashMap<String, OkHttpClient>()
  private val secureKeyAlias = "blinky_remote_credentials"
  private val transferChunkSize = 16 * 1024 * 1024

  override fun definition() = ModuleDefinition {
    Name("BlinkySecureSocket")
    Events("onOpen", "onMessage", "onClose", "onError", "onTransferProgress")

    AsyncFunction("connect") { id: String, url: String, pin: String ->
      connect(id, url, pin)
    }
    AsyncFunction("sendText") { id: String, data: String ->
      check(sockets[id]?.send(data) == true) { "Secure socket could not send the message: $id" }
    }
    AsyncFunction("sendBinary") { id: String, base64Data: String ->
      val bytes = Base64.decode(base64Data, Base64.DEFAULT)
      check(sockets[id]?.send(ByteString.of(*bytes)) == true) {
        "Secure socket could not send the binary message: $id"
      }
    }
    AsyncFunction("close") { id: String ->
      closeSocket(id)
    }
    AsyncFunction("getSecureValue") { key: String ->
      getSecureValue(key)
    }
    AsyncFunction("setSecureValue") { key: String, value: String ->
      setSecureValue(key, value)
    }
    AsyncFunction("deleteSecureValue") { key: String ->
      context.getSharedPreferences("blinky_secure", Context.MODE_PRIVATE)
        .edit().remove(key).apply()
    }
    AsyncFunction("hashFile") { uri: String -> hashFile(uri) }
    AsyncFunction("uploadFile") { options: Map<String, Any?> -> uploadFile(options) }
    AsyncFunction("downloadFile") { options: Map<String, Any?> -> downloadFile(options) }
    OnDestroy {
      sockets.keys.toList().forEach { closeSocket(it) }
    }
  }

  @Synchronized
  private fun connect(id: String, url: String, pin: String) {
    val uri = URI(url)
    require(uri.scheme.equals("wss", ignoreCase = true)) { "Secure socket requires a wss:// URL" }
    require(uri.host != null) { "Secure socket URL must include a hostname" }
    require(pin.startsWith("sha256/")) { "Secure socket requires a sha256/ certificate pin" }

    closeSocket(id)
    val trustManager = PinnedCertificateTrustManager(pin)
    val sslContext = SSLContext.getInstance("TLS").apply {
      init(null, arrayOf<TrustManager>(trustManager), SecureRandom())
    }
    val client = OkHttpClient.Builder()
      // The desktop certificate uses stable localhost SANs while its LAN IP can
      // change. The pinned public key is therefore the server identity check.
      .sslSocketFactory(sslContext.socketFactory, trustManager)
      .hostnameVerifier { _, _ -> true }
      .build()
    val request = Request.Builder().url(url).build()
    clients[id] = client
    val webSocket = client.newWebSocket(request, object : WebSocketListener() {
      override fun onOpen(webSocket: WebSocket, response: Response) {
        synchronized(this@BlinkySecureSocketModule) {
          if (sockets[id] !== webSocket) {
            webSocket.cancel()
            return
          }
          sendEvent("onOpen", mapOf("id" to id))
        }
      }

      override fun onMessage(webSocket: WebSocket, text: String) {
        sendEvent("onMessage", mapOf("id" to id, "data" to text))
      }

      override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
        sendEvent(
          "onMessage",
          mapOf("id" to id, "data" to Base64.encodeToString(bytes.toByteArray(), Base64.NO_WRAP))
        )
      }

      override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
        webSocket.close(code, reason)
      }

      override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        if (removeSocketIfCurrent(id, webSocket)) {
          sendEvent("onClose", mapOf("id" to id))
        }
      }

      override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        if (removeSocketIfCurrent(id, webSocket)) {
          sendEvent("onError", mapOf("id" to id, "error" to (t.message ?: "Secure socket failed")))
        }
      }
    })
    sockets[id] = webSocket
  }

  @Synchronized
  private fun removeSocketIfCurrent(id: String, webSocket: WebSocket): Boolean {
    if (sockets[id] === webSocket) {
      sockets.remove(id)
      clients.remove(id)?.dispatcher?.executorService?.shutdown()
      return true
    }
    return false
  }

  @Synchronized
  private fun closeSocket(id: String) {
    sockets.remove(id)?.cancel()
    clients.remove(id)?.dispatcher?.executorService?.shutdown()
  }

  private fun getSecureValue(key: String): String? {
    val encoded = context.getSharedPreferences("blinky_secure", Context.MODE_PRIVATE)
      .getString(key, null) ?: return null
    return try {
      val packed = Base64.decode(encoded, Base64.NO_WRAP)
      val iv = packed.copyOfRange(0, 12)
      val ciphertext = packed.copyOfRange(12, packed.size)
      val cipher = Cipher.getInstance("AES/GCM/NoPadding")
      cipher.init(Cipher.DECRYPT_MODE, getOrCreateKey(), GCMParameterSpec(128, iv))
      String(cipher.doFinal(ciphertext), Charsets.UTF_8)
    } catch (_: Exception) {
      null
    }
  }

  private fun hashFile(uri: String): Map<String, Any> {
    val digest = MessageDigest.getInstance("SHA-256")
    var size = 0L
    context.contentResolver.openInputStream(Uri.parse(uri))?.use { input ->
      val buffer = ByteArray(1024 * 1024)
      while (true) {
        val count = input.read(buffer)
        if (count < 0) break
        digest.update(buffer, 0, count)
        size += count
      }
    } ?: throw IllegalArgumentException("Unable to open the selected file")
    return mapOf("size" to size, "sha256" to digest.digest().toHex())
  }

  private fun uploadFile(options: Map<String, Any?>): Map<String, Any> {
    val sourceUri = options.string("sourceUri")
    val transferId = options.string("transferId")
    val url = options.string("url")
    val token = options.string("token")
    val pin = options["pin"] as? String
    val totalSize = options.long("size")
    val chunkSize = options.long("chunkSize", transferChunkSize.toLong()).coerceIn(1, transferChunkSize.toLong()).toInt()
    var offset = options.long("offset")
    require(totalSize > 0 && offset in 0..totalSize) { "Invalid upload size or offset" }

    var input = openSourceAtOffset(sourceUri, offset)
    try {
      while (offset < totalSize) {
        val chunk = readStreamChunk(input, minOf(chunkSize.toLong(), totalSize - offset).toInt())
        if (chunk.isEmpty()) throw IllegalStateException("Selected file ended before its reported size")
        val connection = transferConnection(url, pin)
        try {
          connection.requestMethod = "POST"
          connection.doOutput = true
          connection.connectTimeout = 30_000
          connection.readTimeout = 300_000
          connection.setRequestProperty("Authorization", "Bearer $token")
          connection.setRequestProperty("Upload-Offset", offset.toString())
          connection.setRequestProperty("Content-Type", "application/octet-stream")
          connection.setFixedLengthStreamingMode(chunk.size)
          connection.outputStream.use { it.write(chunk) }
          val status = connection.responseCode
          val nextOffset = connection.getHeaderField("Upload-Offset")?.toLongOrNull()
          if (status == 409 && nextOffset != null && nextOffset in 0..totalSize) {
            offset = nextOffset
            input.close()
            input = openSourceAtOffset(sourceUri, offset)
            continue
          }
          if (status != 200 && status != 201) {
            val detail = runCatching { connection.errorStream?.bufferedReader()?.use { it.readText() } }.getOrNull()
            throw IllegalStateException("Upload failed ($status)${if (detail.isNullOrBlank()) "" else ": $detail"}")
          }
          val updated = nextOffset ?: (offset + chunk.size)
          require(updated > offset && updated <= totalSize) { "PC returned an invalid upload offset" }
          offset = updated
          sendEvent("onTransferProgress", mapOf("id" to transferId, "direction" to "upload", "bytes" to offset, "total" to totalSize))
        } finally {
          connection.disconnect()
        }
      }
    } finally {
      input.close()
    }
    return mapOf("transferId" to transferId, "uploadOffset" to offset, "complete" to true)
  }

  private fun openSourceAtOffset(uri: String, offset: Long): java.io.InputStream {
    val input = context.contentResolver.openInputStream(Uri.parse(uri))
      ?: throw IllegalArgumentException("Unable to open the selected file")
    try {
      var remaining = offset
      while (remaining > 0) {
        val skipped = input.skip(remaining)
        if (skipped <= 0) {
          if (input.read() < 0) throw IllegalStateException("Selected file ended before the resume offset")
          remaining -= 1
        } else remaining -= skipped
      }
      return input
    } catch (error: Throwable) {
      input.close()
      throw error
    }
  }

  private fun readStreamChunk(input: java.io.InputStream, maxBytes: Int): ByteArray {
    val buffer = ByteArray(maxBytes)
    var count = 0
    while (count < maxBytes) {
      val read = input.read(buffer, count, maxBytes - count)
      if (read < 0) break
      count += read
    }
    return if (count == buffer.size) buffer else buffer.copyOf(count)
  }

  private fun downloadFile(options: Map<String, Any?>): String {
    val transferId = options.string("transferId")
    val url = options.string("url")
    val token = options.string("token")
    val pin = options["pin"] as? String
    val expectedSize = options.long("size")
    val expectedSha256 = options.string("sha256").lowercase()
    val filename = safeName(options.string("filename"))
    require(expectedSize > 0) { "Invalid download size" }

    val directory = File(context.filesDir, "BlinkyTransfers").apply { mkdirs() }
    val partial = File(directory, "$transferId.part")
    var offset = partial.length()
    if (offset > expectedSize) {
      partial.delete()
      offset = 0
    }
    while (offset < expectedSize) {
      val end = minOf(expectedSize - 1, offset + transferChunkSize - 1)
      val connection = transferConnection(url, pin)
      try {
        connection.requestMethod = "GET"
        connection.connectTimeout = 30_000
        connection.readTimeout = 300_000
        connection.setRequestProperty("Authorization", "Bearer $token")
        connection.setRequestProperty("Range", "bytes=$offset-$end")
        val status = connection.responseCode
        if (status != 206) throw IllegalStateException("Download failed ($status)")
        require(connection.getHeaderField("X-File-Size")?.toLongOrNull() == expectedSize) { "PC returned a different file size" }
        require(connection.getHeaderField("X-File-Sha256")?.equals(expectedSha256, ignoreCase = true) == true) { "PC returned a different file digest" }
        val expectedChunk = end - offset + 1
        var received = 0L
        FileOutputStream(partial, true).use { output ->
          connection.inputStream.use { input ->
            val buffer = ByteArray(1024 * 1024)
            while (true) {
              val count = input.read(buffer)
              if (count < 0) break
              output.write(buffer, 0, count)
              received += count
            }
          }
          output.fd.sync()
        }
        require(received == expectedChunk) { "Download chunk was incomplete" }
        offset += received
        sendEvent("onTransferProgress", mapOf("id" to transferId, "direction" to "download", "bytes" to offset, "total" to expectedSize))
      } finally {
        connection.disconnect()
      }
    }

    val actualHash = hashFile(Uri.fromFile(partial).toString())["sha256"] as String
    require(actualHash.equals(expectedSha256, ignoreCase = true)) { "Downloaded file SHA-256 did not match the PC" }
    val destination = uniqueTransferFile(directory, filename)
    if (!partial.renameTo(destination)) throw IllegalStateException("Could not save the downloaded file")
    return Uri.fromFile(destination).toString()
  }

  private fun transferConnection(url: String, pin: String?): HttpURLConnection {
    val parsed = URI(url)
    require(parsed.scheme.equals("http", true) || parsed.scheme.equals("https", true)) { "Transfer URL must use HTTP or HTTPS" }
    val connection = URL(url).openConnection() as HttpURLConnection
    if (connection is javax.net.ssl.HttpsURLConnection) {
      require(!pin.isNullOrBlank()) { "HTTPS transfers require the saved certificate pin" }
      val trustManager = PinnedCertificateTrustManager(pin)
      val sslContext = SSLContext.getInstance("TLS").apply { init(null, arrayOf<TrustManager>(trustManager), SecureRandom()) }
      connection.sslSocketFactory = sslContext.socketFactory
      connection.hostnameVerifier = javax.net.ssl.HostnameVerifier { _, _ -> true }
    } else {
      require(pin.isNullOrBlank()) { "Pinned release transfers must use HTTPS" }
    }
    return connection
  }

  private fun safeName(value: String): String {
    val cleaned = value.replace(Regex("[\\\\/:*?\"<>|\\p{Cntrl}]"), "_").trim().trim('.')
    return cleaned.takeIf { it.isNotEmpty() && it != ".." } ?: "blinky-download.bin"
  }

  private fun uniqueTransferFile(directory: File, filename: String): File {
    val original = File(directory, filename)
    if (!original.exists()) return original
    val stem = original.nameWithoutExtension
    val extension = original.extension
    for (index in 1..99999) {
      val suffix = if (extension.isEmpty()) " ($index)" else " ($index).$extension"
      val candidate = File(directory, "$stem$suffix")
      if (!candidate.exists()) return candidate
    }
    throw IllegalStateException("Could not find a free file name")
  }

  private fun Map<String, Any?>.string(key: String): String = this[key] as? String
    ?: throw IllegalArgumentException("Missing $key")

  private fun Map<String, Any?>.long(key: String, default: Long = 0): Long = when (val value = this[key]) {
    is Number -> value.toLong()
    null -> default
    else -> throw IllegalArgumentException("Invalid $key")
  }

  private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

  private fun setSecureValue(key: String, value: String) {
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.ENCRYPT_MODE, getOrCreateKey())
    val packed = cipher.iv + cipher.doFinal(value.toByteArray(Charsets.UTF_8))
    context.getSharedPreferences("blinky_secure", Context.MODE_PRIVATE)
      .edit().putString(key, Base64.encodeToString(packed, Base64.NO_WRAP)).apply()
  }

  private fun getOrCreateKey(): SecretKey {
    val keyStore = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
    if (keyStore.containsAlias(secureKeyAlias)) {
      return keyStore.getKey(secureKeyAlias, null) as SecretKey
    }
    val generator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
    generator.init(
      KeyGenParameterSpec.Builder(
        secureKeyAlias,
        KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
      )
        .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
        .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
        .build()
    )
    return generator.generateKey()
  }
}
